msrest/serialization.py

```diff
@@ -137,6 +138,62 @@ def _classify(cls, response, objects):
                 raise DeserializationError("Discriminator {} cannot be absent or null".format(subtype_key))
         return cls
 
+    @classmethod
+    def _solve(cls, dict_data):
+        """Solve this dict against this model a dict where key are attributes.
+
+        This can be used to support case insentive key matching or accept
+        both RestAPI keys and attribute keys in dict input.
+        :param dict dict_data: A dict of data, where keys syntax is uncertain
+        :returns: This same dict with key as intended attributes
+        :raises: KeyError If at least one dict key is not found
+        """
+        attr_keys = {s.lower(): s for s in cls._attribute_map}
+        result_dict = {}
+        for key, value in dict_data.items():
+            lower_key = key.lower()
+            pure_key = attr_keys.pop(lower_key, None)
+            if pure_key:
+                result_dict[pure_key] = value
+            else:
+                for lower_attr_key, attr_key in attr_keys.items():
+                    rest_api_response_key = cls._get_rest_key_part(attr_key)[-1]
+                    if lower_key == rest_api_response_key.lower():
+                        result_dict[attr_key] = value
+                        break
+                else:
+                    raise KeyError("Unable to find attribute for {}".format(key))
+                del attr_keys[lower_attr_key]
+        assert len(result_dict) == len(dict_data)
```

Comment:
Does this handle the case where there are two keys with different casing?  In that case, when lowered, they would stomp on eachother and the size would be different? I'm not sure what the error message looks like in this case, but assertion failures tend not to be super helpful. Could a more descriptive error be given?